File: java/yelp-ai-assistant/src/main/java/com/yelp/assistant/search/StructuredSearchService.java
package com.yelp.assistant.search;

import com.yelp.assistant.model.BusinessData;
import com.yelp.assistant.model.StructuredSearchResult;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class StructuredSearchService {

    private final Map<String, BusinessData> store = new ConcurrentHashMap<>();

    public void addBusiness(BusinessData business) {
        store.put(business.businessId(), business);
    }

    public List<StructuredSearchResult> search(String query, String businessId) {
        BusinessData biz = store.get(businessId);
        if (biz == null) return List.of();

        String q = query.toLowerCase();
        List<String> matched = new ArrayList<>();

        if (containsAny(q, "open", "close", "hour", "time")) matched.add("hours");
        for (String key : biz.amenities().keySet()) {
            String label = key.replace("_", " ");
            if (q.contains(label) || q.contains(key)) matched.add("amenities." + key);
        }
        if (containsAny(q, "address", "location", "where", "phone")) matched.add("address");

        double score = matched.isEmpty() ? 0.5 : 1.0;
        return List.of(new StructuredSearchResult(biz, matched, score));
    }

    private boolean containsAny(String s, String... subs) {
        for (String sub : subs) if (s.contains(sub)) return true;
        return false;
    }
}

File: csharp/yelp-ai-assistant/Search/StructuredSearchService.cs
using System.Collections.Concurrent;
using YelpAiAssistant.Models;

namespace YelpAiAssistant.Search;

public class StructuredSearchService
{
    private readonly ConcurrentDictionary<string, BusinessData> _store = new();

    public void AddBusiness(BusinessData biz) => _store[biz.BusinessId] = biz;

    public IEnumerable<StructuredSearchResult> Search(string query, string businessId)
    {
        if (!_store.TryGetValue(businessId, out var biz)) yield break;
        var q = query.ToLowerInvariant();
        var matched = new List<string>();
        if (ContainsAny(q, "open", "close", "hour", "time")) matched.Add("hours");
        foreach (var key in biz.Amenities.Keys)
        {
            var label = key.Replace("_", " ");
            if (q.Contains(label) || q.Contains(key)) matched.Add("amenities." + key);
        }
        if (ContainsAny(q, "address", "location", "where", "phone")) matched.Add("address");
        yield return new StructuredSearchResult(biz, matched, matched.Count > 0 ? 1.0 : 0.5);
    }

    private static bool ContainsAny(string s, params string[] subs) => subs.Any(s.Contains);
}
